internal/wasm/call_context_test.go

```diff
@@ -288,3 +290,75 @@ func TestCallContext_CallDynamic(t *testing.T) {
 		require.False(t, ok, "expected no opened files")
 	})
 }
+
+func TestCallContext_SetExitCodeOnCanceledOrTimeout(t *testing.T) {
+	t.Run("timeout", func(t *testing.T) {
+		cc := &CallContext{Closed: new(uint64), module: &ModuleInstance{Name: "test"}}
+		const duration = time.Second
+		ctx, cancel := context.WithTimeout(context.Background(), duration)
+		defer cancel()
+		done := cc.SetExitCodeOnCanceledOrTimeout(ctx)
+		time.Sleep(duration * 2)
+		defer done()
+
+		err := cc.FailIfClosed()
+		require.EqualError(t, err, "module \"test\" closed with context deadline exceeded")
+	})
+	t.Run("cancel", func(t *testing.T) {
+		cc := &CallContext{Closed: new(uint64), module: &ModuleInstance{Name: "test"}}
+		ctx, cancel := context.WithCancel(context.Background())
+		done := cc.SetExitCodeOnCanceledOrTimeout(ctx)
+		cancel()
+		defer done()
+
+		time.Sleep(time.Second)
+		err := cc.FailIfClosed()
+		require.EqualError(t, err, "module \"test\" closed with context canceled")
+	})
+
+	t.Run("cancel works", func(t *testing.T) {
```

Comment:
maybe add something for CancelCauseFunc as the cancel error can be arbitrary

Reply:
that is the Go 1.20 feature, so won't add it here yet.